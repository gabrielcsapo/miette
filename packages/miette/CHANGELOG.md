# 0.1.1

- removes unnecessary dependencies and adds API docs.

# 0.1.0

- @gabrielcsapo: improves rendering engine (https://github.com/gabrielcsapo/miette/pull/3)
  - ensures source has a new line on the end if not appends it
  - improves rendering snippet logic
  - allow rendering multiple snippets per line
  - snippets will get a color blind friendly color to distinguish themselves on the line
  - snippet block now closes the block
  - use unicode if supported, but fallback to ascii if not
  - renders link if provided
  - line width is computed correctly instead of always being 0

# 0.0.1

- @gabrielcsapo: basic functionality support and CI set up!
