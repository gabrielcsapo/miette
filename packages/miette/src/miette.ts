import { IError } from "./types";
import { Reporter } from "./reporter";
import { Diagnostic } from "./diagnostic";
import { Theme } from "./theme";

function isUnicodeSupported() {
  if (process.platform !== "win32") {
    return process.env.TERM !== "linux"; // Linux console (kernel)
  }

  return (
    Boolean(process.env.CI) ||
    Boolean(process.env.WT_SESSION) || // Windows Terminal
    process.env.ConEmuTask === "{cmd::Cmder}" || // ConEmu and cmder
    process.env.TERM_PROGRAM === "vscode" ||
    process.env.TERM === "xterm-256color" ||
    process.env.TERM === "alacritty"
  );
}

/**
 * Miette Decorator function to be used on Error
 * @public
 * @param code - error code
 * @param source - raw source that miette needs to annotate
 */
export function miette(code: string, source: string) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/ban-types
  return function <T extends { new (...args: any[]): {} }>(constructor: T) {
    const capturedStack = {};
    return class extends constructor {
      stackHolder = Error.captureStackTrace(capturedStack, this.constructor);

      name = code;
      stack = MietteError({
        // @ts-ignore
        error: this,
        source: source,
        // @ts-ignore
        message: this.message,
        // @ts-ignore
        snippets: this.snippets,
        // @ts-ignore
        diagnostic: this.diagnostic,
      });
    };
  };
}

/**
 * @public
 * @param error - An instance of an error object to be used by the diagnostic reporter
 * @returns string to print to console
 */
export function MietteError(error: IError): string {
  const diagnostic = new Diagnostic(error);
  const reporter = new Reporter(
    diagnostic,
    isUnicodeSupported() ? Theme.unicode().styles : Theme.ascii().styles,
    isUnicodeSupported() ? Theme.unicode().characters : Theme.ascii().characters
  );

  reporter.render();

  return reporter.debugString;
}
