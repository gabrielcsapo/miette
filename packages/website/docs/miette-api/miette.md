---
id: miette
hide_title: true
custom_edit_url: null
title: miette package
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) > [miette](./miette.md)

## miette package

## Classes

|  Class | Description |
|  --- | --- |
|  [Diagnostic](./miette.diagnostic.md) |  |
|  [Reporter](./miette.reporter.md) |  |

## Functions

|  Function | Description |
|  --- | --- |
|  [miette(code, source)](./miette.miette.md) | Miette Decorator function to be used on Error |
|  [MietteError(error)](./miette.mietteerror.md) |  |

## Interfaces

|  Interface | Description |
|  --- | --- |
|  [IDiagnostic](./miette.idiagnostic.md) |  |
|  [IError](./miette.ierror.md) |  |
|  [ILine](./miette.iline.md) |  |
|  [ISnippet](./miette.isnippet.md) |  |
|  [IThemeCharacters](./miette.ithemecharacters.md) |  |
|  [IThemeStyle](./miette.ithemestyle.md) |  |

## Variables

|  Variable | Description |
|  --- | --- |
|  [Theme](./miette.theme.md) |  |
|  [ThemeCharacters](./miette.themecharacters.md) |  |
|  [ThemeStyle](./miette.themestyle.md) |  |

## Type Aliases

|  Type Alias | Description |
|  --- | --- |
|  [Severity](./miette.severity.md) |  |
