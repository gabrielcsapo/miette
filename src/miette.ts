import { IError } from "./types";
import { Reporter } from "./reporter";
import { Diagnostic } from "./diagnostic";
import { Theme } from "./theme";

function isUnicodeSupported() {
  if (process.platform !== "win32") {
    return process.env.TERM !== "linux"; // Linux console (kernel)
  }

  return (
    Boolean(process.env.CI) ||
    Boolean(process.env.WT_SESSION) || // Windows Terminal
    process.env.ConEmuTask === "{cmd::Cmder}" || // ConEmu and cmder
    process.env.TERM_PROGRAM === "vscode" ||
    process.env.TERM === "xterm-256color" ||
    process.env.TERM === "alacritty"
  );
}

export function miette(code: string, source: string) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/ban-types
  return function <T extends { new (...args: any[]): {} }>(constructor: T) {
    const capturedStack = {};
    return class extends constructor {
      stackHolder = Error.captureStackTrace(capturedStack, this.constructor);

      name = code;
      stack = MietteError({
        // @ts-ignore
        error: this,
        source: source,
        // @ts-ignore
        message: this.message,
        // @ts-ignore
        snippets: this.snippets,
        // @ts-ignore
        diagnostic: this.diagnostic,
      });
    };
  };
}

export function MietteError(error: IError): string {
  const diagnostic = new Diagnostic(error);
  const reporter = new Reporter(
    diagnostic,
    isUnicodeSupported() ? Theme.unicode().styles : Theme.ascii().styles,
    isUnicodeSupported() ? Theme.unicode().characters : Theme.ascii().characters
  );

  reporter.render();

  return reporter.debugString;
}
