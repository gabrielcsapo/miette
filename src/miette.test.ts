import prettier from "prettier";

import { miette } from "./miette";

describe("miette", () => {
  test("should be able to render with decorators", () => {
    @miette(
      "foo::bar::baz",
      prettier.format(FooBarBaz.toString(), { parser: "babel" })
    )
    class ShouldBeFalseError extends Error {
      diagnostic = {
        help: "Please consult the guides at http://github.com/foo/bar#guides",
      };

      snippets = [
        {
          context: "if (true)",
          highlight: "This will always be called",
        },
      ];
    }

    function FooBarBaz() {
      // eslint-disable-next-line no-constant-condition
      if (true) {
        try {
          throw new ShouldBeFalseError("Should make things dynamic");
        } catch (error) {
          if (error instanceof Error) {
            console.log(error.stack);
            expect(error.stack).toMatchSnapshot();
          }
        }
      }
    }

    FooBarBaz();
  });
});
