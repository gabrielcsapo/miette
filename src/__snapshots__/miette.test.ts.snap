// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`miette should be able to render with decorators 1`] = `
"Error: foo::bar::baz

   × Should make things dynamic

    ╭───[miette.test.ts:30:17]
 0  │ function FooBarBaz() {
 1  │   // eslint-disable-next-line no-constant-condition
 2  │   if (true) {
    ·     ────┬────
    ·         ╰── This will always be called    
 3  │     try {
 4  │       throw new ShouldBeFalseError(\\"Should make things dynamic\\");
 5  │     } catch (error) {
 6  │       if (error instanceof Error) {
 7  │         console.log(error.stack);
 8  │         expect(error.stack).toMatchSnapshot();
 9  │       }
 10 │     }
 11 │   }
 12 │ }
 13 │ 
    ╰───
‽ Please consult the guides at http://github.com/foo/bar#guides

"
`;
